### PR TITLE
Fix playing a map with NoFail mod never completing

### LINKED ISSUE #1400: HasFailed doesn't take NoFail into account
This causes the onCompletionEvent to never fire.

Reproduce the bug:
Put on NoFail
Have 0 HP left when the song ends.
No screen on completion.

I have a fix for this.

### PR DESCRIPTION
**Pull request by odgaard**
Fixes #1400

Also implements the use of alreadyFailed directly in HasFailed to prevent unnecessary calls to the evaluation functions.

### CODE REVIEW

**Comment by odgaard** [comment]
Any developers want to comment on the error that AppVeyor produces on the PR? Removing the parentheses that AppVeyor suggests would work, but reading it would require knowledge about operator priority, which isn't exactly clean code.

**Comment by peppy** [comment]
Those are the current standards we are complying to, so removing them is correct.

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs [hunk 1]
```diff
@@ -69,7 +69,7 @@ public abstract class ScoreProcessor
         /// <summary>
         /// Whether the score is in a failed state.
         /// </summary>
-        public virtual bool HasFailed => Health.Value == Health.MinValue;
+        public virtual bool HasFailed => alreadyFailed || Health.Value == Health.MinValue && Failed?.Invoke() != false;
```
Comment (peppy) [inline osu.Game/Rulesets/Scoring/ScoreProcessor.cs]: putting this invoke call in a getter property isn't good. the `Failed?.Invoke` part needs to remain in `UpdateFailed`

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs [hunk 2]
```diff
@@ -69,7 +69,12 @@ public abstract class ScoreProcessor
         /// <summary>
         /// Whether the score is in a failed state.
         /// </summary>
-        public virtual bool HasFailed => Health.Value == Health.MinValue;
+        public virtual bool HasFailed => alreadyFailed;
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Scoring/ScoreProcessor.cs]: Think we can remove `alreadyFailed`, and just keep `HasFailed` as a public-get-private-set property?

File: osu.Game/Rulesets/Scoring/ScoreProcessor.cs [hunk 3]
```diff
@@ -69,7 +69,12 @@ public abstract class ScoreProcessor
         /// <summary>
         /// Whether the score is in a failed state.
         /// </summary>
-        public virtual bool HasFailed => Health.Value == Health.MinValue;
+        public virtual bool HasFailed => alreadyFailed;
+
+        /// <summary>
+        /// The conditions for failing
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Scoring/ScoreProcessor.cs]: Missing full-stop.

**Review by smoogipoo — DISMISSED** [review]

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
Looks good 👍

**Pull request merged**